relay-replays/src/recording.rs

```diff
@@ -65,6 +71,121 @@ static STRING_STATE: Lazy<ProcessingState> = Lazy::new(|| {
+}
+
+impl Transform for &'_ mut ScrubberTransform<'_> {
+    fn transform_str<'a>(&mut self, v: &'a str) -> Cow<'a, str> {
+        self.transform_string(v.to_owned())
+    }
+
+    fn transform_string(&mut self, mut value: String) -> Cow<'static, str> {
+        if let Some(ref mut processor) = self.processor1 {
+            if processor
+                .process_string(&mut value, &mut Meta::default(), &STRING_STATE)
+                .is_err()
+            {
+                return Cow::Borrowed("");
+            }
+        }
+
+        if let Some(ref mut processor) = self.processor2 {
+            if processor
+                .process_string(&mut value, &mut Meta::default(), &STRING_STATE)
+                .is_err()
+            {
+                return Cow::Borrowed("");
+            }
+        }
+
+        Cow::Owned(value)
+    }
+}
+
+/// Helper that runs data scrubbing on a raw JSON value during serialization.
+///
+/// This is used by [`EventStreamVisitor`] to serialize recording events on-the-fly from a stream.
+/// It uses a [`ScrubberTransform`] holding all state to perform the actual work.
+struct ScrubbedValue<'a, 'b>(&'a RawValue, Rc<RefCell<ScrubberTransform<'b>>>);
+
+impl serde::Serialize for ScrubbedValue<'_, '_> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut transform = self.1.borrow_mut();
+        let mut deserializer = serde_json::Deserializer::from_str(self.0.get());
+        let scrubber = crate::transform::Deserializer::new(&mut deserializer, &mut *transform);
+        serde_transcode::transcode(scrubber, serializer)
+    }
+}
+
+/// A visitor that deserializes, scrubs, and serializes a stream of recording events.
+struct EventStreamVisitor<'a, S> {
+    serializer: S,
+    scrubber: Rc<RefCell<ScrubberTransform<'a>>>,
+}
+
+impl<'a, S> EventStreamVisitor<'a, S> {
+    fn new(serializer: S, scrubber: Rc<RefCell<ScrubberTransform<'a>>>) -> Self {
+        Self {
+            serializer,
+            scrubber,
+        }
+    }
+}
+
+impl<'de, 'a, S> de::Visitor<'de> for EventStreamVisitor<'a, S>
+where
+    S: ser::Serializer,
+{
+    type Value = S::Ok;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(formatter, "a replay recording event stream")
+    }
+
+    fn visit_seq<A>(self, mut v: A) -> Result<Self::Value, A::Error>
+    where
+        A: de::SeqAccess<'de>,
+    {
+        #[derive(Clone, Copy, serde::Deserialize)]
+        struct TypeHelper {
+            #[serde(rename = "type")]
+            ty: u8,
+        }
+
+        use serde::ser::SerializeSeq;
+        let mut seq = self.serializer.serialize_seq(v.size_hint()).map_err(s2d)?;
+
+        while let Some(raw) = v.next_element::<&'de RawValue>()? {
+            // filter for sentry-specific events
+            let helper = serde_json::from_str::<TypeHelper>(raw.get()).unwrap();
+            if helper.ty == 5 {
```

Comment:
Maybe the const with description what the type is this would be more descriptive here